packages/cozy-client/src/CozyClient.js

```diff
@@ -198,42 +198,54 @@ class CozyClient {
       event: TRIGGER_CREATION
     }
   ) {
-    if (!document._type) return document
-
-    const schema = this.schema.getDoctypeSchema(document._type)
-    if (!schema.cozyMetadata) return document
-
-    const enriched = {
-      ...document
-    }
-    const metadata = Object.entries(schema.cozyMetadata)
-
-    for (const [
-      metadataName,
-      { trigger, value, useCurrentDate = false }
-    ] of metadata) {
-      const shouldSetMetadata =
-        options.event === TRIGGER_CREATION ||
-        (options.event === TRIGGER_UPDATE && trigger === TRIGGER_UPDATE)
-      if (!shouldSetMetadata) continue
-
-      let newValue
-      const metadataPath = `cozyMetadata.${metadataName}`
-
-      const shouldAccumulateValues = Array.isArray(value)
-      if (shouldAccumulateValues) {
-        const prevValue = get(document, metadataPath, [])
-        newValue = uniqBy([...prevValue, ...value])
-      } else if (useCurrentDate) {
-        newValue = new Date().toISOString()
-      } else {
-        newValue = get(document, metadataPath, value)
+    if (!document._type || this.metadata === undefined) return document
+
+    const {
+      doctypeVersion,
+      slug,
+      sourceAccount = null,
+      version
+    } = this.metadata
+
+    const now = new Date().toISOString()
+
+    let cozyMetadata = get(document, 'cozyMetadata', {})
+    if (options.event === TRIGGER_CREATION) {
+      cozyMetadata = {
+        ...cozyMetadata, // custom metadata that are set by the app
+        doctypeVersion,
+        createdByApp: slug,
+        sourceAccount,
+        createdAt: now,
+        createdByAppVersion: version,
+        updatedAt: now,
+        updatedByApps: [
+          {
+            date: now,
+            slug
+          }
+        ]
+      }
+    } else if (options.event === TRIGGER_UPDATE) {
+      cozyMetadata = {
+        // add doctypeVersion and sourceAccount if they don't exist
+        doctypeVersion,
```

Comment:
This one seems dangerous. We assume that updating an existing document will always respect the new schema. Will it?

I need comments from whoever has one.

Reply:
Yeah I wouldn't automatically update that. We currently try very hard to avoid having multiple versions of the same doctype in the wild, but still.

Reply:
Don't try to hard. We *will* have documents in different versions in the future. This will happen once we have third party apps, or when someone is recovering an old export, or when someone play directly with the API.

It's OK to avoid it as much as we can but we definitly should design our apps and libs to be resistant to having documents of different versions in the same doctype.

Reply:
I'm not sure I understand what is the problem.
(However, I was not convinced it was a good idea to create the `doctypeVersion` if it does not exist)

Reply:
You have a contact document which don't have a `version` attribute (so it's probably an old one). You touch it in a small way, you probably don't want to assume it now follows the version=3 correctly

Reply:
Ok I see. Indeed, it's the caller's responsibility to change the metadata in this case.